toolchain/check/subst.cpp

```diff
@@ -0,0 +1,241 @@
+  auto Size() -> int { return worklist.size(); }
+  auto Push(SemIR::InstId inst_id) -> void {
+    worklist.push_back({.inst_id = inst_id,
+                        .is_expanded = false,
+                        .next_index = static_cast<int>(worklist.size() + 1)});
+    CARBON_CHECK(worklist.back().next_index > 0) << "Constant too large.";
+  }
+  auto Pop() -> SemIR::InstId { return worklist.pop_back_val().inst_id; }
+
+ private:
+  // Constants can get pretty large, so use a large worklist. This should be
+  // about 4KiB, which should be small enough to comfortably fit on the stack,
+  // but large enough that it's unlikely that we'll need a heap allocation.
+  llvm::SmallVector<WorklistItem, 512> worklist;
+};
+
+}  // namespace
+
+// Pushes the specified operand onto the worklist.
+static auto PushOperand(Context& context, Worklist& worklist,
+                        SemIR::IdKind kind, int32_t arg) -> void {
+  switch (kind) {
+    case SemIR::IdKind::For<SemIR::InstId>:
+      worklist.Push(static_cast<SemIR::InstId>(arg));
+      break;
+    case SemIR::IdKind::For<SemIR::TypeId>:
+      worklist.Push(context.types().GetInstId(static_cast<SemIR::TypeId>(arg)));
+      break;
+    case SemIR::IdKind::For<SemIR::InstBlockId>:
+      for (auto inst_id :
+           context.inst_blocks().Get(static_cast<SemIR::InstBlockId>(arg))) {
+        worklist.Push(inst_id);
+      }
+      break;
+    case SemIR::IdKind::For<SemIR::TypeBlockId>:
+      for (auto type_id :
+           context.type_blocks().Get(static_cast<SemIR::TypeBlockId>(arg))) {
+        worklist.Push(context.types().GetInstId(type_id));
+      }
+      break;
+    default:
+      break;
+  }
+}
+
+// Converts the operands of this instruction into `InstId`s and pushes them onto
+// the worklist.
+static auto ExpandOperands(Context& context, Worklist& worklist,
+                           SemIR::InstId inst_id) -> void {
+  auto inst = context.insts().Get(inst_id);
+  auto kinds = inst.ArgKinds();
+  PushOperand(context, worklist, kinds.first, inst.arg0());
+  PushOperand(context, worklist, kinds.second, inst.arg1());
+}
+
+// Pops the specified operand from the worklist and returns it.
+static auto PopOperand(Context& context, Worklist& worklist, SemIR::IdKind kind,
+                       int32_t arg) -> int32_t {
+  switch (kind) {
+    case SemIR::IdKind::For<SemIR::InstId>:
+      return worklist.Pop().index;
+    case SemIR::IdKind::For<SemIR::TypeId>:
+      return context.GetTypeIdForTypeInst(worklist.Pop()).index;
+    case SemIR::IdKind::For<SemIR::InstBlockId>: {
+      auto old_inst_block_id = static_cast<SemIR::InstBlockId>(arg);
+      auto size = context.inst_blocks().Get(old_inst_block_id).size();
+      SemIR::CopyOnWriteInstBlock new_inst_block(context.sem_ir(),
+                                                 old_inst_block_id);
+      for (auto i : llvm::index_range(0, size)) {
+        new_inst_block.Set(size - i - 1, worklist.Pop());
+      }
```

Comment:
Can you:

```suggestion
      for (auto i : llvm::reverse(llvm::seq(size))) {
        new_inst_block.Set(i, worklist.Pop());
      }
```

If that works, it'd apply below too.

FWIW, there's probably an option here where you add two APIs to worklist:
- `Back(N)` to return an ArrayRef of the last N items
- `Pop(N)` to pop the last N items

Then do:

```
for (auto [i, inst_id] : llvm::enumerate(llvm::reverse(worklist.Back(N))) {
  new_inst_block.Set(i, N);
}
worklist.Pop(N);
```

But I'm not sure this is really worth it...